circleci: replace Linux tests on Travis CI with CircleCI

aykevl opened the pull request:
End goal: replace Travis CI with CircleCI. CircleCI is much faster and doesn't have an uncertain future like Travis CI.

For now, let's just replace the Linux tests with CircleCI.

johanbrandhorst commented on .circleci/config.yml:
```diff
@@ -0,0 +1,53 @@
+          llvm: "-7"
+      - dep
+      - run: go install .
+      - run: make test
```
Should this not be in a separate `test` job?

aykevl replied:
It seems to me that a new job would mean reinstalling all dependencies, which is rather expensive (at least until we have a special testing container).

johanbrandhorst replied:
Fair point. We might be able to do the pulling in an initial job and then use WORKSPACE to populate for future jobs? https://circleci.com/docs/2.0/caching/.

aykevl replied:
I don't think it's reasonably possible to cache dependencies installed by `apt-get`. What could be improved is to cache submodules or Go dependencies, but I'm not sure how to do that reliably. The Go build cache is already cached, which shaves of ~15s of build time. Improving the TinyGo build cache (for compiler-rt) would also save ~15s, but I'll need better caching logic for that (see buildcache.go).

aykevl commented:
To me, this looks ready to merge.

johanbrandhorst commented:
LGTM, good start! Do we want to add a badge to the README as well?

aykevl commented:
> Do we want to add a badge to the README as well?

Good idea, added right now.

aykevl commented:
It says "NO BUILDS" but that's because this branch hasn't been merged in the dev branch yet.

aykevl commented:
Fixed conflicts. For some reason, they didn't show up in the GitHub UI.

aykevl commented:
Going to experiment a bit...

aykevl commented:
Ok, happy with it. I split up the `apt-get install` line into multiple lines, to make conflicts easier to resolve or avoid them entirely.

deadprogram commented:
I see that all the CircleCI tests now pass. A couple of questions:

Shouldn't this PR be against the `dev` branch?
Should we now be switching to using CircleCI for all builds and removing the Travis build steps?
Should we be using the CircleCi build badge on the README?

aykevl commented:
> Shouldn't this PR be against the `dev` branch?

My bad, fixed now.

> Should we now be switching to using CircleCI for all builds and removing the Travis build steps?

I'd like to, but at the moment we don't have CircleCI testing for macOS. This PR changes the Linux tests, with the eventual goal of changing macOS tests as well. I think they can coexist for the time being.

> Should we be using the CircleCi build badge on the README?

I have added the CircleCI badge to the README. I think we should keep the TravisCI badge for the time being, until we don't use it anymore.

deadprogram commented:
OK, then. Merging into dev, taking that first step.
